tools/PI/DevHome.PI/ViewModels/WinLogsPageViewModel.cs

```diff
@@ -131,69 +196,28 @@ private void TargetApp_PropertyChanged(object? sender, PropertyChangedEventArgs
 
     private void StopWinLogs(bool shouldCleanLogs = true)
     {
-        _winLogsHelper?.Stop();
+        _winLogsService?.Stop();
 
         if (shouldCleanLogs)
         {
             ClearWinLogs();
         }
     }
 
-    private void WinLogsOutput_CollectionChanged(object? sender, NotifyCollectionChangedEventArgs e)
+    private void WinLogEntries_CollectionChanged(object? sender, NotifyCollectionChangedEventArgs e)
     {
-        if (e.Action == NotifyCollectionChangedAction.Add && e.NewItems != null)
+        _dispatcher.TryEnqueue(() =>
         {
-            _dispatcher.TryEnqueue(() =>
+            if (e.Action == NotifyCollectionChangedAction.Add && e.NewItems != null)
             {
                 foreach (WinLogsEntry newEntry in e.NewItems)
                 {
-                    WinLogEntries.Add(newEntry);
                     ThreadPool.QueueUserWorkItem((o) => FindPattern(newEntry.Message));
                 }
-            });
-        }
-    }
-
-    public void Dispose()
-    {
-        _winLogsHelper?.Dispose();
-        GC.SuppressFinalize(this);
-    }
-
-    public void LogStateChanged(object sender, RoutedEventArgs e)
-    {
-        var box = e.OriginalSource as CheckBox;
-        if ((box is not null) && (box.Tag is not null))
-        {
-            var isChecked = box.IsChecked;
-
-            if (_logMeasures)
-            {
-                App.Log("DevHome.PI_WinLogs_LogStateChanged", LogLevel.Measure, new LogStateChangedEventData(box.Name, (box.IsChecked ?? false) ? "true" : "false"), null);
             }
 
-            var tool = (WinLogsTool)box.Tag;
-            _winLogsHelper?.LogStateChanged(tool, isChecked ?? false);
-        }
-    }
-
-    public void UpdateClipboardContent(object sender, DataGridRowClipboardEventArgs e)
-    {
-        var winLogEntry = e.Item as WinLogsEntry;
-        var dataGrid = sender as DataGrid;
-        if ((winLogEntry is not null) && (dataGrid is not null))
-        {
-            // Message Column is the last column.
-            var messageColumnIndex = dataGrid.Columns.Count - 1;
-            var selectedColumnIndex = dataGrid.CurrentColumn.DisplayIndex;
-
-            // Clear clipboard if the selected column is Message column
-            // to copy only the selected text from the textbox.
-            if ((winLogEntry.SelectedText.Length > 0) && (selectedColumnIndex == messageColumnIndex))
-            {
-                e.ClipboardRowContent.Clear();
-            }
-        }
+            UpdateWinLogsViewSource();
```

Comment:
From looking at this, it appears that when a new log comes in, you'll create a new sorted list and assign it to the view. I think this has the unfortunate side effect of causing the UX to scroll to the top of the view.... very painful for the user if they are looking at logs a couple pages in.

If possible, you should try and add items to the existing list instead of creating a new list. This will give the user a much better experience. Check out what I did with the WER stuff to ensure items are inserted in the correct place in the list based on the sort criteria.

Reply:
I had looked at the WER code, however, that works only for sorting. Looking online for sorting and filtering for DataGrids, CollectionViewSource is the recommended way and unfortunately that involved creating a new list every time.

CollectionViewSource makes sure, we don't have to store multiple copies of the ObservableCollection list.

I didn't realize the UX scrolling to top of the view issue, I will work on a fix for it.

Reply:
I found AdvancedCollectionView which helps in sorting and filtering in WinUI3. CollectionViewSource is recommended for WPF, since it has sorting and filtering options in WPF.